library/core/src/iter/range.rs

```diff
@@ -631,32 +731,22 @@ range_incl_exact_iter_impl! {
 impl<A: Step> DoubleEndedIterator for ops::Range<A> {
     #[inline]
     fn next_back(&mut self) -> Option<A> {
-        if self.start < self.end {
-            // SAFETY: just checked precondition
-            self.end = unsafe { Step::backward_unchecked(self.end.clone(), 1) };
-            Some(self.end.clone())
-        } else {
-            None
-        }
+        <Self as RangeIteratorImpl>::specialized_next_back(self)
     }
 
     #[inline]
     fn nth_back(&mut self, n: usize) -> Option<A> {
-        if let Some(minus_n) = Step::backward_checked(self.end.clone(), n) {
-            if minus_n > self.start {
-                // SAFETY: just checked precondition
-                self.end = unsafe { Step::backward_unchecked(minus_n, 1) };
-                return Some(self.end.clone());
-            }
-        }
-
-        self.end = self.start.clone();
-        None
+        <Self as RangeIteratorImpl>::specialized_nth_back(self, n)
     }
 }
 
-#[unstable(feature = "trusted_len", issue = "37572")]
-unsafe impl<A: Step> TrustedLen for ops::Range<A> {}
+macro_rules! impl_trusted_len_for_range {
+    ($($type:ty)*) => {$(
+        #[unstable(feature = "trusted_len", issue = "37572")]
+        unsafe impl TrustedLen for ops::Range<$type> {}
+    )*}
+}
+impl_trusted_len_for_range![char i8 i16 i32 i64 i128 isize u8 u16 u32 u64 u128 usize];
```

Comment:
What's the relation between `TrustedLen` and `TrustedStep`? Would `unsafe impl TrustedLen for Range<T> where T: TrustedStep {}` work or is there a case where it can be `TrustedStep` but not `TrustedLen`?

Reply:
This will likely go in the docs when I get to it. Upon initial thought, such an impl would be plausible. I'd have to think about the invariants and whether it makes sense for such an impl to exist.

Basically is `TrustedLen` means that `Iterator::size_hint` is guaranteed to be correct, while `TrustedStep` means that the various methods of `Step` uphold the invariants such that the `unsafe` methods can be used internally (rather than panicking when they are not).

Reply:
The blanket implementations `impl<T: TrustedStep> TrustedLen for core::ops::Range<T>` and `impl<T: TrustedStep> TrustedLen for core::ops::RangeFrom<T>` are be sound. Specifically, the following invariants for `Step::steps_between` exist:

> * `steps_between(&a, &b) == Some(n)` only if `a <= b`
>    Note that `a <= b` does _not_ imply `steps_between(&a, &b) != None`;
>    this is the case when it would require more than `usize::MAX` steps to get to `b`
> * `steps_between(&a, &b) == None` if `a > b`

The first invariant is what is generally required for `TrustedLen` to be sound. The note addendum satisfies an additional `TrustedLen` invariant.

> The upper bound must only be `None` if the actual iterator length is larger than `usize::MAX`

The second invariant logically follows the first so long as the `PartialOrd` implementation is correct; regardless it is explicitly stated. If `a < b` then `(0, Some(0))` is returned by `ops::Range<A: Step>::size_hint`. As such the second invariant is upheld.